use dts for mmu mapping

JiafeiPan opened the pull request:
devicetree: add mmu mapping support

   Currently for ARM64, mmu mapping entries are hardcoded in soc's
   mmu_regions.c, and such mapping will be used for all board and
   applications although some mapping are not useful for some
   applications.

   So this patch provides another method to add mmu mapping entry
   by devicetree, so such each application can specify mmu mapping
   by devicetree overlay.

   Devicetree need to do the following update:
    1. Add property "zephyr,memory-region" to specify mmu entry name.
    2. Add property "zephyr,memory-region-mmu" to specify mmu entry
       attribute with string array which is valid c token from  architecture
        speficied MMU header file.
    3. The address and size will from "reg" node.

    For example:

            ccm: ccm@30380000 {
                    compatible = "nxp,imx-ccm";
                    reg = <0x30380000 DT_SIZE_K(64)>;
                    zephyr,memory-region = "CCM";
                    zephyr,memory-region-mmu = "MT_DEVICE_nGnRnE", "MT_P_RW_U_NA", "MT_NS";
                    #clock-cells = <3>;
           };

    And then in C code to include mmu entries specified in devicetree:
    For example in "mmu_regions.c" for ARM64 platform
    ...
    ...

    static const struct arm_mmu_region mmu_regions[] = {
            ...
            LINKER_DT_REGION_MMU(MMU_REGION_FLAT_ENTRY)
            ...
    };

With this dts example, the code generated will be:
```
   static const struct arm_mpu_region mpu_regions[] = {
       ...
       { .name = "CCM",
         .base = 0x30380000,
         .size = DT_SIZE_K(64),
         .attr = MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS, },
       ...
   };
```

JiafeiPan committed:
devicetree: add mmu mapping support

Currently for ARM64, mmu mapping entries are hardcoded in soc's
mmu_regions.c, and such mapping will be used for all board and
applications although some mapping are not useful for some
applications.

So this patch provides another method to add mmu mapping entry
by devicetree, so such each application can specify mmu mapping
in devicetree overlay.

Devicetree need to do the following update:
1. Add property "zephyr,memory-region" to specify mmu entry name.
2. Add property "zephyr,memory-region-mmu" to specify mmu entry
   attribute with string array which is valid c token from
   architecture speficied MMU header file.
3. The address and size will from "reg" node.

For example:

ccm: ccm@30380000 {
   compatible = "nxp,imx-ccm";
   reg = <0x30380000 DT_SIZE_K(64)>;
   zephyr,memory-region = "CCM";
   zephyr,memory-region-mmu = "MT_DEVICE_nGnRnE", "MT_P_RW_U_NA", "MT_NS";
   #clock-cells = <3>;
};

And then in C code include mmu entries specified in devicetree:
For example in "mmu_regions.c" for ARM64 platform
...
...

static const struct arm_mmu_region mmu_regions[] = {
	...
	LINKER_DT_REGION_MMU(MMU_REGION_FLAT_ENTRY)
	...
};

Another update is to move "zephyr,memory-region" and define
""zephyr,memory-region-mmu" in base.yaml, so that each device node with
"reg" property can be used together with these two properties to add
mmu entries without modify specific device tree yaml file.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
board: mimx8m/n/p: use dts for mmu mapping

Move MMU mapping for UART4 from mmu_region.c to dts.
Removed UART2 mmu mapping as it is not used by any board.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
The similar design for MMU with https://github.com/zephyrproject-rtos/zephyr/pull/43119/files

carlocaione commented:
@JiafeiPan did you mean to close this PR?

JiafeiPan commented:
> @JiafeiPan did you mean to close this PR?

Hi, @carlocaione , my initial idea is to leverage compatible "zephyr,memory-region" to figure out the dts node which need to do mmu mapping, so I move zephyr,memory-region.yaml to be included in base.yaml, but I found with this PR, there is some error

JiafeiPan committed:
scripts: gen_defines: define DT_FOREACH_OKAY_VARGS_HELPER

Export DT_FOREACH_OKAY_VARGS_HELPER macro, it has function arguments
compared with DT_FOREACH_OKAY_HELPER.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
devicetree: add DT_FOREACH_STATUS_OKAY_NODE_VARGS macro

"fn" in this macro can take extra parameters compared to macro
DT_FOREACH_STATUS_OKAY_NODE.

Added test coverage in tests/lib/devicetree/api/src/main.c.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
> > @JiafeiPan did you mean to close this PR?
> 
> Hi, @carlocaione , my initial idea is to leverage compatible "zephyr,memory-region" to figure out the dts node which need to do mmu mapping, so I move zephyr,memory-region.yaml to be included in base.yaml, but I found with this PR, there is some error for ARM32 platform, so I want to change the design to only to do mmu mapping the dts node with property of "zephyr,memory-region-mmu" but not use compatible string "zephyr,memory-region", but currently I can't find any API which is like "DT_FOREACH_STATUS_OKAY_WITH_PROPERTY(property_string, FN, ...)", any suggestions? thanks.

Hi, @carlocaione , I worked out a new implementation, it will don't use "zephyr,memory-region"  compatible, only need to check the node: 
1. status okay
2. both with property: "zephyr,memory-region",       "zephyr,memory-region-mmu",  "reg"

Then add mmu mapping entry for these dts nodes. please help to review, thanks.

JiafeiPan commented:
Dear maintainers, this PR is ready for review, please help to review, and I would like to merge  it into Zephyr v3.2 release,  thanks.

carlocaione commented:
What's the use case here?

The problem I have with this PR is that we should try to avoid and discourage the usage of `mmu_regions.c` entirely. That was a quick and dirty way to be able to have a 1:1 (direct) mapping in the MMU when the MMU driver was in its infancy. Now the correct way for a driver to have its MMIO space mapped is by using the Device MMIO API.

In your example the CCM driver should be using the Device MMIO API to map its own space without having to resort to extend the `mmu_regions.c` at all. If an application doesn't need that driver (and its MMU mapping) then it can simply remove it from the DT (and the mapping will be gone with it).

The only reason `mmu_regions.c` is still there is because the GIC driver is not a proper driver and cannot use the Device MMIO APIs (yet).

So, from a technical prospective this PR is fine but I'm worried this will basically enable the bad habit of using `mmu_regions.c` instead of modifying the drivers to properly support the Device MMIO APIs.


See #46477 for an extensive discussion on this topic.

@povergoing @SgrrZhf what do you think about this?

povergoing commented:
> Now the correct way for a driver to have its MMIO space mapped is by using the Device MMIO API.

Agreed, 1) it should be the drivers' responsibility to do the MMIO mapping and then do some configs or something important. 2) using DTS for mapping seems to increase the complexity of device tree usage. I doubt it could be too many config entries in the DTS if we apply this for a complicated device like PCI + AHCI

JiafeiPan commented:
I agree with that driver should take responsibility to do the MMIO mapping.

But Carlo said "the big issue with Zephyr is that 95% of the drivers are not using the device MMIO API",  especially in some cases user maybe use hal driver directly which is not a standard Zephyr driver, and in general some hal driver will be shared with other RTOS, so not possible to modify hal driver to add MMIO mapping.

Another concern is some hal driver is using physical address directly, if we don't do 1:1 mapping, hal driver should has memory access issues.

So my suggestions is we need to ask drivers to support MMIO mapping if it is possible, in the other way, provide user another easier method to do 1:1 mapping.

Thanks.

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -2092,6 +2092,20 @@
+ * @param fn macro to invoke
+ * @param ... variable number of arguments to pass to fn
+ */
+#define DT_FOREACH_STATUS_OKAY_VARGS_NODE(fn, ...) DT_FOREACH_OKAY_VARGS_HELPER(fn, __VA_ARGS__)
```
1. Rename this to DT_FOREACH_STATUS_OKAY_NODE_VARGS. It is the "VARGS" version of DT_FOREACH_STATUS_OKAY_NODE. It is not the "NODE" version of DT_FOREACH_STATUS_OKAY_VARGS.
2. Add test coverage in tests/lib/devicetree/api/src/main.c

JiafeiPan replied:
okay, will update it.

mbolivar-nordic commented on dts/bindings/base/base.yaml:
```diff
@@ -109,3 +109,21 @@ properties:
       required: false
       type: string-array
       description: Provided names of mailbox / IPM channel specifiers
+
+    zephyr,memory-region:
+      type: string
+      required: false
+      description: |
+        Signify that this node should result in a dedicated linker script
+        memory region in the final executable. The region address and size
+        is taken from the <reg> property, while the name is the value of
+        this property.
+
+    zephyr,memory-region-mmu:
+      type: string-array
+      required: false
+      description: |
+        Signify that this node should result in a dedicated MMU region. The
+        region address and size are taken from the <reg> property, while the MMU
+        attribute is the value of this string array property, each attribute
+        string will be a valid C token.
```
Whoa, hold on, why are these going into base.yaml? The commit log doesn't explain. Adding base.yaml properties needs a good reason.

JiafeiPan replied:
The idea is that these two prosperities can be used in any node which has "reg" property,  as "reg" is is base.yaml, so I think we also need to put these two together with "reg",  otherwise if they are put in some standalone yaml file, for example, put them into dts/bindings/base/zephyr,memory-region.yaml where current "zephyr, memory-region" locates, we have to include this ymal file in each device driver node yaml files, I don't think it is a better choice, so I suggest put it base.yaml, so that no need update the other yaml files anymore.

mbolivar-nordic commented on scripts/dts/gen_defines.py:
```diff
@@ -862,6 +862,9 @@ def write_global_macros(edt):
     out_dt_define("FOREACH_OKAY_HELPER(fn)",
                   " ".join(f"fn(DT_{node.z_path_id})" for node in edt.nodes
                            if node.status == "okay"))
+    out_dt_define("FOREACH_OKAY_VARGS_HELPER(fn, ...)",
```
You must update doc/guides/dts/macros.bnf whenever you add new generated macros.

JiafeiPan replied:
okay, will update it.

mbolivar-nordic commented on include/zephyr/linker/devicetree_regions.h:
```diff
@@ -187,6 +205,29 @@
+ * and call the EXPAND_MMU_FN() macro
+ */
+#define _CHECK_MMU_ATTR_FN(node_id, EXPAND_MMU_FN, ...)					\
+	COND_CODE_1(UTIL_AND_AND(DT_NODE_HAS_PROP(node_id, zephyr_memory_region_mmu),	\
```
Is it really worth adding UTIL_AND_AND just to use it here? Not a blocking comment since I don't maintain the util header, but it doesn't seem justified since you can just nest COND_CODE_1 here. I think it is risky to start adding macros like UTIL_AND_AND -- where do you stop? UTIL_AND_OR? UTIL_OR_AND? UTIL_OR_OR? UTIL_AND_AND_AND? ...

JiafeiPan replied:
Okay, will remove this macro.

carlocaione commented:
> especially in some cases user maybe use hal driver directly which is not a standard Zephyr driver, and in general some hal driver will be shared with other RTOS, so not possible to modify hal driver to add MMIO mapping.

> Another concern is some hal driver is using physical address directly, if we don't do 1:1 mapping, hal driver should has memory access issues.

Those are valid concerns, still an hard pillow to swallow TBH.

carlocaione commented on dts/bindings/base/base.yaml:
```diff
@@ -109,3 +109,21 @@ properties:
+      type: string
+      required: false
+      description: |
+        Signify that this node should result in a dedicated linker script
+        memory region in the final executable. The region address and size
```
This is actually only true for architectures supporting this. ARM64 is not one of those. Can you expand this with:

`Signify that this node should result in a dedicated linker script memory region in the final executable when supported by the linker script`.

JiafeiPan replied:
Updated, but replace linker with build.

carlocaione commented on include/zephyr/linker/devicetree_regions.h:
```diff
@@ -187,6 +205,30 @@
+ *
+ * and call the EXPAND_MMU_FN() macro
+ */
+#define _CHECK_MMU_ATTR_FN(node_id, EXPAND_MMU_FN, ...)						\
```
Can you rename the current `_CHECK_ATTR_FN` to `_CHECK_MPU_ATTR_FN` for coherence as well?

JiafeiPan replied:
Done.

carlocaione commented on include/zephyr/linker/devicetree_regions.h:
```diff
@@ -187,6 +205,30 @@
+ * Call _CHECK_MMU_ATTR_FN() for each enabled node passing EXPAND_MMU_FN() as
+ * explicit argument and the user-provided MMU_FN() macro in __VA_ARGS__
+ */
+#define _CHECK_APPLY_MMU_FN(EXPAND_MMU_FN, ...)					\
```
same here, can you rename `_CHECK_APPLY_FN` to `_CHECK_APPLY_MPU_FN`?

carlocaione commented on include/zephyr/linker/devicetree_regions.h:
```diff
@@ -290,4 +332,78 @@
+ *
+ * Example devicetree fragment:
+ *
+ *     / {
+ *             soc {
+ *                     ccm: ccm@30380000 {
+ *                         compatible = "nxp,imx-ccm";
+ *                         reg = <0x30380000 DT_SIZE_K(64)>;
+ *                         zephyr,memory-region = "CCM";
+ *                         zephyr,memory-region-mmu = "MT_DEVICE_nGnRnE", "MT_P_RW_U_NA", "MT_NS";
+ *                     };
+ *             };
+ *     };
```
This between `@code{.dts}` and `@endcode`

While you are at it, can you fix lines 250-257 in the same file?

JiafeiPan replied:
Done.

JiafeiPan replied:
Done.

carlocaione commented on include/zephyr/linker/devicetree_regions.h:
```diff
@@ -290,4 +332,78 @@
+ *   };
+ *
+ */
+#define LINKER_DT_REGION_MMU(mmu_fn) _CHECK_APPLY_MMU_FN(_EXPAND_MMU_FN, mmu_fn)
```
At this point I'm realizing how misnamed are `LINKER_DT_REGION_MPU` and this macro  because these macros have nothing to do with the linker script.

Can you rename those to something like `BUILD_DT_REGION_{MPU,MMU}`?

JiafeiPan replied:
Done.

carlocaione commented on soc/arm64/nxp_imx/mimx8m/mmu_regions.c:
```diff
@@ -25,25 +26,17 @@ static const struct arm_mmu_region mmu_regions[] = {
 			      DT_REG_SIZE(DT_NODELABEL(iomuxc)),
 			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
+
+	LINKER_DT_REGION_MMU(MMU_REGION_FLAT_ENTRY)
```
This should actually be added to all the `mmu_regions.c` files for the ARM64 arch.

JiafeiPan replied:
Done

carlocaione reviewed (changes requested):
Besides the comments left, if we decide to have this merged we should at lest convert all the current `mmu_regions.c` files to use this new mechanism and at least having the GIC entries removed from there.

SgrrZhf commented:
> Now the correct way for a driver to have its MMIO space mapped is by using the Device MMIO API.

I think so. If we really want a 1:1 mapping, maybe we could introduce a feature in Device MMIO API that does 1:1 mapping. It's that possible?

SgrrZhf commented:
Understood, those memory pages must be 1:1 mapped before running some HAL driver.

JiafeiPan committed:
soc: arm64: add BUILD_DT_REGION_MMU in all mmu_regions.c

Add BUILD_DT_REGION_MMU for all ARM64 SoC to leverage dts node
properity to add mmu mapping entries.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
devicetree: change LINKER_DT_REGION_MPU to BUILD_DT_REGION_MPU

It is more meaningful as the script is handled in build process.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan commented:
Dear maintainers, any more comments or can we merge this PR by today? thanks.

carlocaione commented:
> Dear maintainers, any more comments or can we merge this PR by today? thanks.

At least @mbolivar-nordic needs to ACK this before merging

mbolivar-nordic reviewed (changes requested):
I  think either there should be a good reason why all of these properties belong in base.yaml, or they all should continue to live in zephyr,memory-region.yaml. The rest of the DT API changes look good, though, thank you.

galak commented:
Need to read through the thread, but this really feels like it should be in the driver and NOT in devicetree.

JiafeiPan commented:
> I think either there should be a good reason why all of these properties belong in base.yaml, or they all should continue to live in zephyr,memory-region.yaml. The rest of the DT API changes look good, though, thank you.

Hi, @mbolivar-nordic, The idea is that these two properties can be used in any node which has "reg" property, currently "reg" is is base.yaml, so I think we'd better put these two together with "reg", otherwise if they are put in some standalone yaml file, for example, put them into dts/bindings/base/zephyr,memory-region.yaml where current "zephyr, memory-region" locates, we have to include this ymal file in each device driver node yaml files which want to use this feature, there will be more modifications.

JiafeiPan commented:
> I think either there should be a good reason why all of these properties belong in base.yaml, or they all should continue to live in zephyr,memory-region.yaml. The rest of the DT API changes look good, though, thank you.

Hi, @mbolivar-nordic, The idea is that these two properties can be used in any node which has "reg" property, currently "reg" is is base.yaml, so I think we'd better put these two together with "reg", otherwise if they are put in some standalone yaml file, for example, put them into dts/bindings/base/zephyr,memory-region.yaml where current "zephyr, memory-region" locates, we have to include this ymal file in each device driver node yaml files which want to use this feature, there will be more modifications.

mbolivar-nordic commented:
OK, I'm going to leave my -1 on until @galak has a chance to come back to this since he has misgivings as well. I will let you sort it out with him. I have no MMU experience (I'm a microcontroller and userspace sort of engineer!), so it sounds best for me to avoid making this discussion noisier.

JiafeiPan commented:
> Need to read through the thread, but this really feels like it should be in the driver and NOT in devicetree.

Hi, @galak , not sure whether have go through this thread, and what's your comments? thanks.

carlocaione commented:
@JiafeiPan if you feel it is valuable you can bring this to the next dev review meeting.

galak commented:
I'm against this change.  As @carlocaione stated, this should be handled via the MMIO api and having drivers doing the mapping.  Introducing this properties in the dts is NOT the right solution.

galak reviewed (changes requested):
I'm against this change.  As @carlocaione stated, this should be handled via the MMIO api and having drivers doing the mapping.  Introducing this properties in the dts is NOT the right solution.

JiafeiPan commented:
Sorry for the late response because of holiday.

Yes, I agree that for native Zephyr driver, it should handle memory mapping.

Maybe this new feature is only useful for the use-case that application call hal driver APIs directly and don't use native Zephyr drivers, and hal driver mostly likely don't handle memory mapping.

Anyway, I keep my opinions and depend on maintainers decision whether to merge it or not, thanks.
